src/ApiCore/AuthWrapper.php

```diff
@@ -57,15 +60,49 @@ public function __construct(FetchAuthTokenInterface $fetchAuthTokenInterface, ca
      */
     public function getBearerString()
     {
-        return 'Bearer ' . $this->getToken();
+        return 'Bearer ' . self::getToken($this->fetchAuthTokenInterface, $this->authHttpHandler);
+    }
+
+    /**
+     * @return callable Callable function that returns an authorization header.
+     */
+    public function getAuthorizationHeaderCallback()
+    {
+        $fetchAuthTokenInterface = $this->fetchAuthTokenInterface;
+        $authHttpHandler = $this->authHttpHandler;
+
+        // NOTE: changes to this function should be treated carefully and tested thoroughly. It will
+        // be passed into the gRPC c extension, and changes have the potential to trigger very
+        // difficult-to-diagnose segmentation faults.
+        return function () use ($fetchAuthTokenInterface, $authHttpHandler) {
+            return ['authorization' => ['Bearer ' . self::getToken($fetchAuthTokenInterface, $authHttpHandler)]];
+        };
     }
 
-    private function getToken()
+    private static function getToken($fetchAuthTokenInterface, $authHttpHandler)
     {
-        $token = $this->fetchAuthTokenInterface->getLastReceivedToken();
-        if (is_null($token) || time() > $token['expires_at']) {
-            $token = $this->fetchAuthTokenInterface->fetchAuthToken($this->authHttpHandler);
+        $token = $fetchAuthTokenInterface->getLastReceivedToken();
+        if (self::isExpired($token)) {
+            $token = $fetchAuthTokenInterface->fetchAuthToken($authHttpHandler);
         }
-        return $token['access_token'];
+        if (self::isValid($token)) {
```

Comment:
Since `isExpired` checks validity as well, WDYT about moving this check into the condition above?

```php
$token = $fetchAuthTokenInterface->getLastReceivedToken();
if (self::isExpired($token)) {
    $token = $fetchAuthTokenInterface->fetchAuthToken($authHttpHandler);
    if (!self::isValid($token)) {
         return '';
    }
}

return $token['access_token'];
```

Reply:
The case I wanted to try to catch was if `$token` was invalid, then avoid trying to access the array with `$token['access_token']`. In turned out not to be the case, but I was wondering if trying to do an array access on a non-array object in this function (but called from the c extension) might have been causing the segfault.

Rearranging in that way means that if `getLastReceivedToken()` returns an invalid value, we will still try to perform array access.

Reply:
I thought `isExpired` checks validity, so in that case wouldn't it fall through and attempt to fetch a fresh token?

Reply:
Ah yes, you are totally right. Fixed.